### ISSUE TITLE
Retry not working, no consumers listed in rabbitmq admin dashboard

### ISSUE THREAD

**Issue by mchristensen76** [guidesmiths/rascal#34]
I need help figuring out this issue.  I use rascal with multiple microservices and it works really well.  But I have one microservice that after some amount of time is no longer connected to rabbitmq.  I go to the admin dashboard for rabbitmq and see no consumers listed for the queue used by this microservice.  In this last instance I see a lot of this error in the logs:

```
Unexpected close
Channel ended, no reply will be forthcoming
```

I don't know if that's related but I see it many times, so that makes me think it was able to reconnect.    Any idea's on things I can look at to help figure this out?  Do you know of any cases where this might happen?

**Comment by cressie176** [comment]
Hi @mchristensen76, sorry you're having problems.

* What version of rascal are you using?
* Can you paste the relevant rascal code showing any error handlers and rascal config?
* Are you able to recreate the issue reliably? If so, can you run with DEBUG='rascal:*'?

**Comment by cressie176** [comment]
I doubt the error is related to the recovery config (this is for when you nack a message). I suspect what's happening is that there's a [connection error](https://github.com/squaremo/amqp.node/blob/6ee18f27e31863f5256bc921a21f36ebf2ba5e74/lib/connection.js#L271-L272) (maybe due to inactivity if it happens over night - I've seen firewalls / VPNs close connections for this reasons unless configured not to).

I don't have a theory yet as to what's causing ```Channel ended, no reply will be forthcoming``` yet though. This is reported by amqplib when a channel is closed while a reply was outstanding. The replies are generated by any [channel](https://github.com/squaremo/amqp.node/blob/13741dd362eee1be7bef75067af68fb3b0c6fd36/lib/callback_model.js) that involves an rpc.

Hopefully the rascal debug will shed some light

**Comment by mchristensen76** [comment]
@cressie176 here's some more debug info:

This "The channel has been closed.  Unable to ack message" keeps repeating:
```
2018-11-09T10:52:30.755Z rascal:Subscription Received message: 4a9f24e2-0043-4614-abfa-c0495e2a73d3 from queue: aria.healthcheck_incident_tracker
2018-11-09T10:52:30.755Z rascal:SubscriberError Handling subscriber error for message: 4a9f24e2-0043-4614-abfa-c0495e2a73d3
2018-11-09T10:52:30.755Z rascal:SubscriberError Attempting to recover message: 4a9f24e2-0043-4614-abfa-c0495e2a73d3 using strategy: nack
2018-11-09T10:52:30.756Z rascal:SubscriberSession Not acknowledging message: 4a9f24e2-0043-4614-abfa-c0495e2a73d3 with requeue: false on channel: 2a2ef043-6950-4ff1-adb9-07d3d47bd31d
2018-11-09T10:52:30.757Z rascal:SubscriberError Message: 4a9f24e2-0043-4614-abfa-c0495e2a73d3 was recovered using stragegy: nack
2018-11-09T10:52:30.758Z rascal:SubscriberError Message: 4a9f24e2-0043-4614-abfa-c0495e2a73d3 was republished to queue: aria.healthcheck_incident_tracker 1 times
2018-11-09T10:52:30.758Z rascal:SubscriberError Message: 4a9f24e2-0043-4614-abfa-c0495e2a73d3 failed to be recovered using stragegy: republish
[2018-11-09T10:52:30.758Z] error: 	The channel has been closed. Unable to ack message```

Then from there the following error starts to repeat:
```
2018-11-09T10:52:39.253Z rascal:Vhost Error creating channel: 062653a4-20ee-4b14-8551-293e0590d0fb from amqps://admin:***@redacted/redacted?heartbeat=1&connection_timeout=10000&channelMax=100: No channels left to allocate
2018-11-09T10:52:39.253Z rascal:SubscriberError Message: e470b907-2a47-4efd-a44a-e95e80d87783 failed to be recovered using stragegy: republish```

You also see this message mixed in:
```2018-11-09T10:59:19.209Z rascal:Subscription Handling channel error: 							Unexpected close from aria.default.subscription using channel: 9267dcd5-1671-4882-96ae-07da93dbdd9c
2018-11-09T10:59:19.209Z rascal:Subscription Subscribing to queue: aria.healthcheck_incident_tracker
2018-11-09T10:59:19.209Z rascal:Vhost Requested channel. Outstanding channel requests: 25```

**Comment by cressie176** [comment]
Thanks for the debug. I'll try to work out what's going on.

**Comment by cressie176** [comment]
I suspect there might be multiple things going wrong and we might need to address one by one. My first observation is that the message `4a9f24e2-0043-4614-abfa-c0495e2a73d3` is attempting to be recovered twice.

1. Message: 4a9f24e2-0043-4614-abfa-c0495e2a73d3 was recovered using stragegy: nack
1. Message: 4a9f24e2-0043-4614-abfa-c0495e2a73d3 failed to be recovered using stragegy: republish

Your config doesn't include a recovery strategy that starts with `nack`. Can you confirm that...

* The `forward` wasn't missing from the [debug](https://github.com/guidesmiths/rascal/blob/master/lib/amqp/SubscriberError.js#L97)
* Other than in the `deferred_retry` recovery strategy you never use the `nack` recovery strategy, either in config or inline.

**Comment by cressie176** [comment]
Can you also show me all your rascal config (except usernames + passwords)

**Comment by cressie176** [comment]
FYI I've published version 3.2.2 of rascal. I have a theory that your subscriber is getting an invalid message (e.g. the content type is set to application/json, but the message content is not parsable). It would explain where the `nack` recovery strategy was coming from, but not much else. It would be great if you could update.

**Comment by mchristensen76** [comment]
I'll try the update soon, thx for getting that done.

Here's the config for reference:

```
rascalConfig: {
    "vhosts": {
        "redacted": {
            "connections": [
                {
                    "slashes": true,
                    "protocol": "amqps",
                    "user": "redacted",
                    "password": "redacted",
                    "hostname": "redacted",
                    "port": "redacted",
                    "vhost": "redacted",
                    "options": {
                        "heartbeat": 1
                    },
                    "socketOptions": {
                        "timeout": 1000,
                        "servername": "redacted"
                    }
                },
                {
                    "slashes": true,
                    "protocol": "amqps",
                    "user": "redacted",
                    "password": "redacted",
                    "hostname": "redacted",
                    "port": "redacted",
                    "vhost": "redacted",
                    "options": {
                        "heartbeat": 1
                    },
                    "socketOptions": {
                        "timeout": 1000,
                        "servername": "redacted"
                    }
                }
            ],
            "exchanges": [
                "aria.event_hub",
                "aria.delay",
                "aria.retry",
                "aria.dead_letters"
            ],
            "queues": {
                "aria.delay:1m": {
                    "options": {
                        "arguments": {
                            "x-message-ttl": 60000,
                            "x-dead-letter-exchange": "aria.retry"
                        }
                    }
                },
                "aria.patch_incident_tracker": {
                    "options": {
                        "arguments": {
                            "x-max-priority": 10,
                            "x-dead-letter-exchange": "aria.dead_letters",
                            "x-dead-letter-routing-key": "aria.patch_incident_tracker.dead_letter"
                        }
                    }
                },
                "dead_letters.aria.patch_incident_tracker": {}
            },
            "bindings": {
                "aria.delay[aria.delay.1m] -> aria.delay:1m": {},
                "aria.retry[redacted] -> aria.patch_incident_tracker": {},
                "aria.event_hub[redacted] -> aria.patch_incident_tracker": {},
                "aria.dead_letters[aria.patch_incident_tracker.dead_letter] -> dead_letters.aria.patch_incident_tracker": {}
            },
            "subscriptions": {
                "aria.default.subscription": {
                    "queue": "aria.patch_incident_tracker",
                    "contentType": "application/json",
                    "redeliveries": {
                        "limit": 5,
                        "counter": "shared"
                    }
                }
            },
            "publications": {
                "retry_in_1m": {
                    "exchange": "aria.delay",
                    "options": {
                        "CC": [
                            "aria.delay.1m"
                        ]
                    }
                },
                "my_q": {
                    "queue": "aria.patch_incident_tracker",
                    "encryption": "aria.crypto.default"
                },
                "aria.event_hub": {
                    "exchange": "aria.event_hub",
                    "encryption": "aria.crypto.default"
                }
            }
        }
    },
    "encryption": {
        "aria.crypto.default": {
            "key": "redacted",
            "ivLength": 16,
            "algorithm": "aes-256-cbc"
        }
    },
    "recovery": {
        "deferred_retry": [
            {
                "strategy": "forward",
                "attempts": 10,
                "publication": "retry_in_1m",
                "xDeathFix": true
            },
            {
                "strategy": "nack"
            }
        ],
        "dead_letter": [
            {
                "strategy": "republish",
                "immediateNack": true
            }
        ]
    },
    "redeliveries": {
        "counters": {
            "shared": {
                "size": 1000,
                "type": "inMemory"
            }
        }
    },
    "defaults": {
        "vhosts": {
            "publicationChannelPools": {
                "regularPoolSize": 1,
                "confirmPoolSize": 1
            },
            "connection": {
                "slashes": true,
                "protocol": "amqp",
                "hostname": "localhost",
                "user": "guest",
                "password": "guest",
                "port": "5672",
                "options": {
                    "heartbeat": 10,
                    "connection_timeout": 10000,
                    "channelMax": 100
                },
                "retry": {
                    "min": 1000,
                    "max": 60000,
                    "factor": 2,
                    "strategy": "exponential"
                },
                "socketOptions": {
                    "timeout": 10000
                }
            },
            "exchanges": {
                "assert": true,
                "type": "topic"
            },
            "queues": {
                "assert": true
            },
            "bindings": {
                "destinationType": "queue",
                "bindingKey": "#"
            }
        },
        "publications": {
            "vhost": "/",
            "confirm": true,
            "options": {
                "persistent": true,
                "mandatory": true
            }
        },
        "subscriptions": {
            "vhost": "/",
            "prefetch": 10,
            "retry": {
                "min": 1000,
                "max": 60000,
                "factor": 2,
                "strategy": "exponential"
            },
            "redeliveries": {
                "limit": 100,
                "timeout": 1000,
                "counter": "stub"
            },
            "deferCloseChannel": 60000
        },
        "redeliveries": {
            "counters": {
                "stub": {},
                "inMemory": {
                    "size": 1000
                }
            }
        }
    }
}
```

**Comment by cressie176** [comment]
Wow. I never actually imagine anyone else would work out how to use Rascal in that depth! It all looks good. Couple of observations, which will have no bearing on this particular problem and I suspect you've already considered

1. the inMemory counter is only useful if the poison messages don't crash your process. We've started using https://github.com/cressie176/rascal-redis-counter
1. I wonder if a heartbeat of 1s is quite short (but might be appropriate in your environment).

Another question - what version of node are you using?

**Comment by cressie176** [comment]
Any update?

**Comment by mchristensen76** [comment]
I'm using node v10.11.0.

I updated to the latest rascal version and will see how it goes.

**Comment by mchristensen76** [comment]
@cressie176 I figured out the cause of the issue.  There was a section of code that would do something like this:

1. Get json docs from database matching criteria.
2. Fetch data from api.
3. Loop through the fetched results and if the value exists in #1, remove it from #1.
3a. Use lodash.findIndex on an object with multiple fields to filter on.

So with that ^ snippet in mind I think what happened is the CPU usage jumped because of lodash.findIndex.  At this point I don't know if the high CPU usage caused the connection loss to rabbitmq and then rascal retried, then the cycle repeats.  At some point rascal stops trying to reconnect, not sure if thats because of a backoff or rabbitmq connection limit.  Or could the event loop have been blocked?  I'm not sure but it was repeatable so I used the inspect command to help me narrow down the problem.

I've since fixed that section of code to be more efficient and haven't had any issues since.

**Comment by cressie176** [comment]
Hi @mchristensen76 

I'm glad you got things working. High CPU usage can block the event loop, which will mean amqplib cannot send heartbeat messages. Consequently the server can close the connection.

This explains why rascal would have retried. It never gives up, but does backoff exponentially. I'll try reproducing and check that rascal behaves sensibly under the High CPU conditions.

Thanks for updating.

**Comment by cressie176** [comment]
I tested by setting up a subscription then kicking off a tight loop for a high number of iterations. After a few seconds the connection was closed by RabbitMQ due to heartbeat failure. Because the tight loop blocked any other operations nothing was seen on the client. When the tight loop eventually finished rascal intercepted the connection close event, re-established the connection and resubscribed.

Closing as there's not much more Rascal can do.
